rpc.go

```diff
@@ -245,12 +247,14 @@ func (l *Libvirt) stream(e event) error {
 	l.em.Unlock()
 
 	if ok {
-		// we close the channel in deregister() so that we don't block here
-		// forever without a receiver. If that happens, this write will panic.
-		defer func() {
-			recover()
+		go func() {
+			// we close the channel in deregister() so that we don't block here
+			// forever without a receiver. If that happens, this write will panic.
+			defer func() {
+				recover()
+			}()
+			c.Events <- e
```

Comment:
In your scenario where stream is opened for a long time, if we receive multiple stream packets while channel is blocked, we'll loose their ordering since order of execution for goroutines is not guaranteed in any way.
For example (found no stackoverflow) in this case:
```
	messages := make(chan string)
	go func() { messages <- "hello" }()
	go func() { messages <- "ping" }()
	msg := <-messages
	msg2 := <-messages
	fmt.Println(msg)
	fmt.Println(msg2)
```
(playground at https://play.golang.org/p/OU29KfWpgNs) always prints "ping" before "hello" for me.
I would guess that even if you add more data to TestVolumeUploadDownloadIntegration test and add some Sleep in the middle, you'll get data in messed up order with this patch.

It looks like we'd need to implement some sort of queueing mechanism here. Maybe add a small buffer to the stream channel and/or just reject packets if stream channel is not ready.

Reply:
This is true - since the new code spawns a goroutine for every response we get back from the client, it could deliver those responses out of-of-order.

Reply:
@YorikSar  you are right about the order problems. But it seems adding a small buffer just decrease the chance of deadlock, not solve it. Since if caller side consumes events slowly, new coming event could still consume all the buffer and trigger deadlock.  What if we add some coordinating mechanisms to gurantee the order of events?